### ISSUE TITLE
RCTModuleData deadlock when running with debugger

### ISSUE THREAD

**Issue by bruzenak** [facebook/react-native#11196]
### Description

When running with the debugger in 0.38, RCTModuleData deadlocks on line 88 of RCTModuleData.mm (std::unique_lock<std::mutex> lock(_instanceLock);). This reproduces reliably with my setup.

From a layman's read, it appears that a module I'm using (react-native-search-bar) is set up to load off the main thread. While it's doing this and grabbing RCTAccessibilityManager, the main thread reloads and tries to grab the lock. Meanwhile RCTAccessibilityManager wants the main thread and already HAS the lock. So it's waiting on dispatch to the main thread, and the main thread is waiting on the accessibility manager to load.

Without the debugger, this doesn't lock. It appears the debugger attaches and triggers a reload in [RCTDevMenu setExecutorClass:], and that reload is what's causing this lock.

### Reproduction

I attached a package.json file that reproduces this issue.

1. Run react-native init BugSample
1a. Drop in package.json
1b. Run npm i
2. Open iOS project in XCode
3. Hit run
4. Attach debugger (Cmd+D)
5. Quit
6. Restart

This project reliably deadlocks for me.

[package.json.zip](https://github.com/facebook/react-native/files/620059/package.json.zip)

### Solution

Don't grab the locks; wait on the full load process to complete before attaching the debugger, etc. Something this close to core should probably be fixed by an engineer that fully understands the module loading stack.

### Additional Information

* React Native version: 0.38.0
* Platform: iOS (only tested on iOS)
* Operating System: MacOS

**Comment by lacker** [comment]
That sounds like it would be a fairly disruptive change, to not grab the locks here. Is there any simpler solution? I am afraid that the cost of fixing this would be above the cost of leaving it as is, unless a bunch more people show up with this issue.

**Comment by bruzenak** [comment]
Well, it's definitely a deadlock. I'm sure there's a better fix than not grabbing the locks there, but it requires a deeper knowledge of how the library loading works than I have at the moment. For example: requiring that the lock gets grabbed before a reload can be triggered might fix it. 

Certainly the current behavior is bad in that it does nothing to prevent deadlocking between two threads. The cost of not fixing it would be that this deadlocks in other situations that don't require the debugger, which I'm better are possible. This is a very severe bug and I wouldn't bet that there aren't other ways to encounter it.

**Comment by Luzgan** [comment]
Well "unless a bunch more people show up with this issue" - I cannot ignore think like that...

> 2017-01-31 12:55:40.643 [warn][tid:com.facebook.react.RCTBridgeQueue][RCTModuleData.mm:220] RCTBridge required dispatch_sync to load RCTAccessibilityManager. This may lead to deadlocks
> 2017-01-31 12:55:44.705 [fatal][tid:com.facebook.react.RCTBridgeQueue] Runtime is not ready for debugging.

Basically it is disallowing me to work with debugger at all, because for me it is happening when I click Debug JS and application tries to reload.

Sadly because of time I won't be able to give much more information in nearest future, just wanted to say something to not make this forgotten. I am using RN0.40.

**Comment by honki91** [comment]
I have the same deadlock issue when trying to use [react-native-youtube ](https://github.com/inProgress-team/react-native-youtube). I'm using RN 0.41.2

**Comment by lacker** [comment]
Since all the bug repros involve specific libraries, maybe it would be more effective to report this issue to the makers of react-native-search-bar and react-native-youtube?

**Comment by bruzenak** [comment]
When I dug into this, it was not a client library bug, it was a React Native core bug. I imagine you'll see a variety of libraries trigger it based on hitting the race condition in the code. I don't think there's anything those libraries could do to fix it except putting in random arbitrary waits.

**Comment by lacker** [comment]
Hmm. I believe you, I am just wondering how to make it simpler for people to be able to debug this. Is there some way to reproduce this without any third-party modules?

**Comment by bruzenak** [comment]
You could probably build a module that exhibited the behavior as part of a project. I'd dig in further but am unfortunately strapped for time. Race conditions suck. Maybe if you built some code to purge the module cache and reloaded a few times?

**Comment by kh1990** [comment]
Update: I noticed the deadlock only occurs for me when I am working on a branch. Each time I submit a pull request and merge into the master, the deadlock is resolved. Maybe this will help some of you.

**Comment by kevando** [comment]
No idea what I'm doing, but I get this same error when I enable the chrome debugger and use this lib

https://github.com/gre/react-native-view-shot

:/

**Comment by ds300** [comment]
I'm also getting this with react-native-search-bar. Anybody found a workaround?

**Comment by rkretzschmar** [comment]
Hi all,
for me - the following solved the problem (deadlock):
I patched `RCTModuleData.mm` by replacing the mutex in `setUpInstanceAndBridge` in a way that it only protects the write access (the `new` operation in this case), rather than all the block - so the method looks like that now:

```obj-c
- (void)setUpInstanceAndBridge
{
  RCT_PROFILE_BEGIN_EVENT(RCTProfileTagAlways, @"[RCTModuleData setUpInstanceAndBridge] [_instanceLock lock]", @{ @"moduleClass": NSStringFromClass(_moduleClass) });
  
  if (!_setupComplete && _bridge.valid) {
    if (!_instance) {
      if (RCT_DEBUG && _requiresMainQueueSetup) {
        RCTAssertMainQueue();
      }
      
      // Only the write access (the new operation in this case) has to be protected really.
      {
        std::unique_lock<std::mutex> lock(_instanceLock);
        
        // This IF-statement is still necessary, because the instance could already be instantiated while this thread was waiting
        if (!_instance) {
          RCT_PROFILE_BEGIN_EVENT(RCTProfileTagAlways, @"[RCTModuleData setUpInstanceAndBridge] [_moduleClass new]",  @{ @"moduleClass": NSStringFromClass(_moduleClass) });
          _instance = [_moduleClass new];
          RCT_PROFILE_END_EVENT(RCTProfileTagAlways, @"");
        }
      }
      
      if (!_instance) {
        // Module init returned nil, probably because automatic instantatiation
        // of the module is not supported, and it is supposed to be passed in to
        // the bridge constructor. Mark setup complete to avoid doing more work.
        _setupComplete = YES;
        RCTLogWarn(@"The module %@ is returning nil from its constructor. You "
                   "may need to instantiate it yourself and pass it into the "
                   "bridge.", _moduleClass);
      }
    }
    
    if (_instance && RCTProfileIsProfiling()) {
      RCTProfileHookInstance(_instance);
    }
    
    // Bridge must be set before methodQueue is set up, as methodQueue
    // initialization requires it (View Managers get their queue by calling
    // self.bridge.uiManager.methodQueue)
    [self setBridgeForInstance];
  }
  
  [self setUpMethodQueue];
  
  RCT_PROFILE_END_EVENT(RCTProfileTagAlways, @"");
  
  // This is called outside of the lock in order to prevent deadlock issues
  // because the logic in `finishSetupForInstance` can cause
  // `moduleData.instance` to be accessed re-entrantly.
  if (_bridge.moduleSetupComplete) {
    [self finishSetupForInstance];
  } else {
    // If we're here, then the module is completely initialized,
    // except for what finishSetupForInstance does.  When the instance
    // method is called after moduleSetupComplete,
    // finishSetupForInstance will run.  If _requiresMainQueueSetup
    // is true, getting the instance will block waiting for the main
    // thread, which could take a while if the main thread is busy
    // (I've seen 50ms in testing).  So we clear that flag, since
    // nothing in finishSetupForInstance needs to be run on the main
    // thread.
    _requiresMainQueueSetup = NO;
  }
}
```

I know that mutex is not the root cause and a deadlock can still be happen, because of all those `dispatch_sync` (`RCTUnsafeExecuteOnMainQueueSync`) calls. But it may at least decreases the chance of a deadlock.

Please - can somebody verify that?

Thanks in advance.

**Comment by rkretzschmar** [comment]
Maybe `[self setUpMethodQueue]` and `[self setBridgeForInstance];` have to be protected separately as well. Haven't thought this through completely. Just mentioning.

**Comment by vikeri** [comment]
@rkretzschmar If you can generate a patch file then I'll happily try it :)

**Comment by rkretzschmar** [comment]
Hi @vikeri - thanks! 
Here is the commit https://github.com/rkretzschmar/react-native/commit/d0a9c7df51b4567b1d11b394a73a1da9eb1c6d85 based on the latest react-native master. Maybe you can just cherry-pick.

**Comment by vikeri** [comment]
@rkretzschmar Awesome! Thanks, I'll try it asap

**Comment by joonhocho** [comment]
I am having the same issues. React native v0.42.3
```
2017-03-31 04:38:31.533 [warn][tid:com.facebook.react.RCTBridgeQueue][RCTModuleData.mm:285] Required dispatch_sync to load constants for RCTSourceCode. This may lead to deadlocks
```

**Comment by rkretzschmar** [comment]
Hi @joonhocho - sorry to hear that. Do you think you have time to try out my patch https://github.com/rkretzschmar/react-native/commit/d0a9c7df51b4567b1d11b394a73a1da9eb1c6d85 ?

**Comment by mikesurowiec** [comment]
This problem had been plaguing me for awhile, I was finally able to fix it by re-ordering the Build Phases section. I moved `libRNSearchBar.a` to the bottom of my build phases and now I can reload JS with the debugger running. Hope this helps someone!

**Comment by xuetongiqn** [comment]
@mikesurowiec  it works for me, Thank you!

**Comment by grabbou** [comment]
Still happens, had to remove that module.

**Comment by hramos** [comment]
Hi there! This issue is being closed because it has been inactive for a while. Maybe the issue has been fixed in a recent release, or perhaps it is not affecting a lot of people. Either way, we're automatically closing issues after a period of inactivity. Please do not take it personally!

If you think this issue should definitely remain open, please let us know. The following information is helpful when it comes to determining if the issue should be re-opened:

- Does the issue still reproduce on the latest release candidate? Post a comment with the version you tested.
- If so, is there any information missing from the bug report? Post a comment with all the information required by the [issue template](https://github.com/facebook/react-native/blob/master/.github/ISSUE_TEMPLATE.md).
- Is there a pull request that addresses this issue? Post a comment with the PR number so we can follow up.

If you would like to work on a patch to fix the issue, *contributions are very welcome*! Read through the [contribution guide](http://facebook.github.io/react-native/docs/contributing.html), and feel free to hop into [#react-native](https://discordapp.com/invite/0ZcbPKXt5bZjGY5n) if you need help planning your contribution.

**Comment by waifo** [comment]
when i am running a new project the simulator works fine, but after some time on reloading the simulator screen turns black and the installed app closes.
I am not able to figure out the real issue.

**Comment by simlmx** [comment]
Unlike @mikesurowiec I didn't have any `libRNSearchBar.a` library but moving some other third-party libraries to the end fixed it for me.
